akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -394,6 +396,62 @@ final case class Fold[In, Out](zero: Out, f: (Out, In) ⇒ Out) extends GraphSta
 /**
  * INTERNAL API
  */
+final class FoldAsync[In, Out](zero: ⇒ Out, f: (Out, In) ⇒ Future[Out]) extends GraphStage[FlowShape[In, Out]] {
+  import akka.dispatch.ExecutionContexts
+
+  val in = Inlet[In]("FoldAsync.in")
+  val out = Outlet[Out]("FoldAsync.out")
+
+  val shape = FlowShape.of(in, out)
+
+  override def toString: String = "FoldAsync"
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+      private var aggregated: Out = zero
+      private var aggregating: Future[Out] = Future.successful(aggregated)
+
+      private def handle(output: Try[Out]): Unit = output match {
+        case Failure(reason) ⇒ fail(out, reason)
+        case Success(o) ⇒
+          aggregated = o
+          if (!isClosed(in)) pull(in)
```

Comment:
If you come here with a success but the upstream is closed the stage should complete